// @shoojs/auth/dist/index.js
const encoder = new TextEncoder();
const PKCE_MAX_AGE_MS = 10 * 60 * 1000;
const DEFAULT_SHOO_BASE_URL = "https://shoo.dev";
const DEFAULT_SESSION_MONITOR_INTERVAL_MS = 60_000;
function assertNoClientSecretOption(value, apiName) {
    if (!value || typeof value !== "object" || !("clientSecret" in value)) {
        return;
    }
    const clientSecret = value.clientSecret;
    if (clientSecret === undefined) {
        return;
    }
    throw new Error(`${apiName} no longer accepts clientSecret in browser code. Use a server-side token exchange.`);
}
function isBrowser() {
    return typeof window !== "undefined";
}
function requireBrowser(api) {
    if (!isBrowser()) {
        throw new Error(`${api} must be called in a browser environment`);
    }
}
function randomString(length = 64) {
    const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";
    const random = crypto.getRandomValues(new Uint8Array(length));
    let value = "";
    for (let i = 0; i < random.length; i += 1) {
        value += chars[random[i] % chars.length];
    }
    return value;
}
function bytesToBase64Url(bytes) {
    let binary = "";
    for (const byte of bytes) {
        binary += String.fromCharCode(byte);
    }
    return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}
function decodeBase64Url(value) {
    const normalized = value.replace(/-/g, "+").replace(/_/g, "/");
    const padded = normalized.padEnd(Math.ceil(normalized.length / 4) * 4, "=");
    if (typeof atob !== "function") {
        throw new Error("atob is not available in this runtime");
    }
    return atob(padded);
}
function parseJson(value) {
    try {
        return JSON.parse(value);
    }
    catch {
        return null;
    }
}
function currentRoute() {
    requireBrowser("currentRoute");
    return `${window.location.pathname}${window.location.search}${window.location.hash}`;
}
function normalizeReturnTo(value) {
    if (!value) {
        return null;
    }
    if (!isBrowser()) {
        return value.startsWith("/") && !value.startsWith("//") ? value : null;
    }
    try {
        const parsed = new URL(value, window.location.origin);
        if (parsed.origin !== window.location.origin) {
            return null;
        }
        const route = `${parsed.pathname}${parsed.search}${parsed.hash}`;
        if (!route.startsWith("/") || route.startsWith("//")) {
            return null;
        }
        return route;
    }
    catch {
        return null;
    }
}
function deriveRedirectUri(callbackPath) {
    requireBrowser("deriveRedirectUri");
    const safePath = normalizeReturnTo(callbackPath) ?? "/auth/callback";
    return new URL(safePath, window.location.origin).toString();
}
export function deriveClientIdFromRedirectUri(redirectUri) {
    return `origin:${new URL(redirectUri).origin}`;
}
function resolveOptions(options) {
    const callbackPath = normalizeReturnTo(options.callbackPath) ?? "/auth/callback";
    const fallbackPath = normalizeReturnTo(options.fallbackPath) ?? "/";
    const redirectUri = options.redirectUri ?? deriveRedirectUri(callbackPath);
    const clientId = options.clientId ?? deriveClientIdFromRedirectUri(redirectUri);
    return {
        shooBaseUrl: options.shooBaseUrl ?? "https://shoo.dev",
        callbackPath,
        redirectUri,
        clientId,
        requestPii: options.requestPii ?? false,
        storageKey: options.storageKey ?? "shoo_identity",
        pkceStorageKey: options.pkceStorageKey ?? "shoo_pkce",
        returnToStorageKey: options.returnToStorageKey ?? "shoo_return_to",
        fallbackPath
    };
}
export async function createPkceBundle() {
    const verifier = randomString(64);
    const state = randomString(32);
    const digest = await crypto.subtle.digest("SHA-256", encoder.encode(verifier));
    const challenge = bytesToBase64Url(new Uint8Array(digest));
    return { state, verifier, challenge };
}
export function createSignInUrl(params) {
    const url = new URL("/authorize", params.shooBaseUrl);
    const effectiveClientId = params.clientId ?? deriveClientIdFromRedirectUri(params.redirectUri);
    url.searchParams.set("client_id", effectiveClientId);
    url.searchParams.set("redirect_uri", params.redirectUri);
    url.searchParams.set("state", params.state);
    url.searchParams.set("code_challenge", params.codeChallenge);
    url.searchParams.set("code_challenge_method", "S256");
    if (params.requestPii) {
        url.searchParams.set("pii", "true");
    }
    return url.toString();
}
export function parseCallback(url = isBrowser() ? window.location.href : "") {
    if (!url) {
        return null;
    }
    const parsed = new URL(url);
    const code = parsed.searchParams.get("code");
    const state = parsed.searchParams.get("state");
    if (!code || !state) {
        return null;
    }
    return { code, state };
}
export function clearCallbackParams(url = isBrowser() ? window.location.href : "") {
    if (!url) {
        return "";
    }
    const next = new URL(url);
    next.searchParams.delete("code");
    next.searchParams.delete("state");
    next.searchParams.delete("error");
    if (isBrowser()) {
        window.history.replaceState({}, "", next.toString());
    }
    return next.toString();
}
export function getIdentity(storageKey = "shoo_identity") {
    requireBrowser("getIdentity");
    const raw = window.localStorage.getItem(storageKey);
    if (!raw) {
        return { userId: null };
    }
    const parsed = parseJson(raw);
    if (!parsed) {
        return { userId: null };
    }
    return {
        userId: parsed.userId ?? parsed.pairwiseSub ?? null,
        token: parsed.token,
        expiresIn: parsed.expiresIn,
        receivedAt: parsed.receivedAt
    };
}
export function persistIdentity(userId, token, storageKey = "shoo_identity", extras) {
    requireBrowser("persistIdentity");
    window.localStorage.setItem(storageKey, JSON.stringify({
        userId,
        token,
        expiresIn: extras?.expiresIn,
        receivedAt: Date.now()
    }));
}
export function clearIdentity(storageKey = "shoo_identity") {
    requireBrowser("clearIdentity");
    window.localStorage.removeItem(storageKey);
}
export function decodeIdentityClaims(idToken) {
    if (!idToken) {
        return null;
    }
    const parts = idToken.split(".");
    if (parts.length < 2) {
        return null;
    }
    const decoded = decodeBase64Url(parts[1]);
    return parseJson(decoded);
}
export async function exchangeCode(params) {
    assertNoClientSecretOption(params, "exchangeCode");
    const tokenEndpoint = new URL("/token", params.shooBaseUrl).toString();
    const body = new URLSearchParams({
        grant_type: "authorization_code",
        client_id: params.clientId,
        redirect_uri: params.redirectUri,
        code: params.code,
        code_verifier: params.codeVerifier
    });
    const response = await fetch(tokenEndpoint, {
        method: "POST",
        headers: {
            "Content-Type": "application/x-www-form-urlencoded"
        },
        body
    });
    if (!response.ok) {
        const details = await response.text();
        throw new Error(`Token exchange failed (${response.status}): ${details || "no details"}`);
    }
    return (await response.json());
}
function readReturnTo(key) {
    requireBrowser("readReturnTo");
    return normalizeReturnTo(window.sessionStorage.getItem(key));
}
function popReturnTo(key) {
    const value = readReturnTo(key);
    window.sessionStorage.removeItem(key);
    return value;
}
function storeReturnTo(key, value, callbackPath, fallbackPath) {
    requireBrowser("storeReturnTo");
    let normalized = normalizeReturnTo(value) ?? fallbackPath;
    if (normalized === callbackPath) {
        normalized = fallbackPath;
    }
    window.sessionStorage.setItem(key, normalized);
    return normalized;
}
function resolveReturnTarget(options, storedReturnTo, defaults) {
    const explicit = normalizeReturnTo(options.redirectTo) ?? normalizeReturnTo(options.returnTo);
    if (explicit) {
        return explicit;
    }
    if (storedReturnTo) {
        return storedReturnTo;
    }
    return normalizeReturnTo(options.fallbackPath) ?? defaults.fallbackPath;
}
function toError(error) {
    return error instanceof Error ? error : new Error("Unexpected session check failure");
}
function resolveClientIdForCheck(params) {
    if (params.clientId) {
        return params.clientId;
    }
    if (params.redirectUri) {
        return deriveClientIdFromRedirectUri(params.redirectUri);
    }
    return null;
}
export async function checkSession(params = {}) {
    requireBrowser("checkSession");
    const shooBaseUrl = params.shooBaseUrl ?? DEFAULT_SHOO_BASE_URL;
    const token = params.token ?? getIdentity(params.storageKey).token;
    if (!token) {
        return {
            status: "login_required",
            reason: "invalid_token"
        };
    }
    const expectedClientId = resolveClientIdForCheck(params);
    if (expectedClientId) {
        const claims = decodeIdentityClaims(token);
        if (!claims || claims.aud !== expectedClientId) {
            return {
                status: "login_required",
                reason: "invalid_token"
            };
        }
    }
    const endpoint = new URL("/session/check", shooBaseUrl).toString();
    const response = await fetch(endpoint, {
        method: "POST",
        headers: {
            Authorization: `Bearer ${token}`
        }
    });
    if (response.status === 404 || response.status === 501) {
        return { status: "unsupported" };
    }
    if (response.status === 200) {
        const payload = (await response.json());
        if (payload.status === "active") {
            return { status: "active" };
        }
        throw new Error("Session check returned an invalid success payload");
    }
    if (response.status === 401) {
        const payload = (await response.json());
        if (payload.status === "login_required" && typeof payload.reason === "string") {
            if (payload.reason === "revoked" || payload.reason === "expired" || payload.reason === "invalid_token") {
                return {
                    status: "login_required",
                    reason: payload.reason
                };
            }
        }
        return {
            status: "login_required",
            reason: "invalid_token"
        };
    }
    throw new Error(`Session check failed (${response.status})`);
}
export function startSessionMonitor(options = {}) {
    requireBrowser("startSessionMonitor");
    const intervalMs = Math.max(1_000, Math.floor(options.intervalMs ?? DEFAULT_SESSION_MONITOR_INTERVAL_MS));
    const immediate = options.immediate !== false;
    let stopped = false;
    let inFlight = false;
    const stop = () => {
        if (stopped) {
            return;
        }
        stopped = true;
        window.clearInterval(timer);
    };
    const runCheck = async () => {
        if (stopped || inFlight) {
            return;
        }
        inFlight = true;
        try {
            const result = await checkSession(options);
            if (result.status === "login_required") {
                options.onLoginRequired?.(result);
                stop();
            }
        }
        catch (error) {
            options.onError?.(toError(error));
        }
        finally {
            inFlight = false;
        }
    };
    const timer = window.setInterval(() => {
        void runCheck();
    }, intervalMs);
    if (immediate) {
        void runCheck();
    }
    return { stop };
}
export function createShooAuth(options) {
    const resolved = resolveOptions(options);
    const clientGetIdentity = (storageKey) => getIdentity(storageKey ?? resolved.storageKey);
    const clientCheckSession = async (params) => {
        const effectiveRedirectUri = params?.redirectUri ?? resolved.redirectUri;
        const effectiveClientId = params?.clientId ?? deriveClientIdFromRedirectUri(effectiveRedirectUri);
        return checkSession({
            ...params,
            shooBaseUrl: params?.shooBaseUrl ?? resolved.shooBaseUrl,
            redirectUri: effectiveRedirectUri,
            clientId: effectiveClientId,
            storageKey: params?.storageKey ?? resolved.storageKey
        });
    };
    const clientStartSessionMonitor = (options) => {
        const effectiveRedirectUri = options?.redirectUri ?? resolved.redirectUri;
        const effectiveClientId = options?.clientId ?? deriveClientIdFromRedirectUri(effectiveRedirectUri);
        return startSessionMonitor({
            ...options,
            shooBaseUrl: options?.shooBaseUrl ?? resolved.shooBaseUrl,
            redirectUri: effectiveRedirectUri,
            clientId: effectiveClientId,
            storageKey: options?.storageKey ?? resolved.storageKey
        });
    };
    const clientStartSignIn = async (params) => {
        requireBrowser("startSignIn");
        const effectiveRedirectUri = params?.redirectUri ?? resolved.redirectUri;
        const effectiveClientId = params?.clientId ?? deriveClientIdFromRedirectUri(effectiveRedirectUri);
        const bundle = await createPkceBundle();
        window.sessionStorage.setItem(resolved.pkceStorageKey, JSON.stringify({
            state: bundle.state,
            verifier: bundle.verifier,
            createdAt: Date.now()
        }));
        storeReturnTo(resolved.returnToStorageKey, params?.returnTo ?? currentRoute(), resolved.callbackPath, resolved.fallbackPath);
        const url = createSignInUrl({
            shooBaseUrl: params?.shooBaseUrl ?? resolved.shooBaseUrl,
            redirectUri: effectiveRedirectUri,
            clientId: effectiveClientId,
            state: bundle.state,
            codeChallenge: bundle.challenge,
            requestPii: params?.requestPii ?? resolved.requestPii
        });
        window.location.assign(url);
        return { url, bundle };
    };
    const clientFinishSignIn = async (params) => {
        requireBrowser("finishSignIn");
        assertNoClientSecretOption(params, "finishSignIn");
        const callback = parseCallback(params?.url ?? window.location.href);
        if (!callback) {
            return null;
        }
        const rawPkce = window.sessionStorage.getItem(resolved.pkceStorageKey);
        if (!rawPkce) {
            throw new Error("Missing PKCE verifier in sessionStorage");
        }
        const parsedPkce = parseJson(rawPkce);
        if (!parsedPkce?.state || !parsedPkce?.verifier) {
            throw new Error("Invalid PKCE verifier payload");
        }
        if (typeof parsedPkce.createdAt === "number" && Date.now() - parsedPkce.createdAt > PKCE_MAX_AGE_MS) {
            window.sessionStorage.removeItem(resolved.pkceStorageKey);
            throw new Error("PKCE verifier expired. Start sign-in again.");
        }
        if (parsedPkce.state !== callback.state) {
            throw new Error("State mismatch in callback");
        }
        const storedReturnTo = params?.consumeReturnTo === false
            ? readReturnTo(resolved.returnToStorageKey)
            : popReturnTo(resolved.returnToStorageKey);
        const effectiveRedirectUri = params?.redirectUri ?? resolved.redirectUri;
        const effectiveClientId = params?.clientId ?? deriveClientIdFromRedirectUri(effectiveRedirectUri);
        const token = await exchangeCode({
            shooBaseUrl: params?.shooBaseUrl ?? resolved.shooBaseUrl,
            clientId: effectiveClientId,
            redirectUri: effectiveRedirectUri,
            code: callback.code,
            codeVerifier: parsedPkce.verifier
        });
        persistIdentity(token.pairwise_sub, token.id_token, resolved.storageKey, {
            expiresIn: token.expires_in
        });
        window.sessionStorage.removeItem(resolved.pkceStorageKey);
        if (params?.redirectAfter) {
            const target = resolveReturnTarget({
                redirectTo: params.redirectTo,
                returnTo: params.returnTo,
                fallbackPath: params.fallbackPath
            }, storedReturnTo, resolved);
            window.location.replace(target);
            return token;
        }
        if (params?.clearCallbackParams !== false) {
            clearCallbackParams();
        }
        return token;
    };
    const clientHandleCallback = async (params) => {
        requireBrowser("handleCallback");
        assertNoClientSecretOption(params, "handleCallback");
        const token = await clientFinishSignIn({
            url: params?.url,
            redirectAfter: false,
            consumeReturnTo: false
        });
        if (!token) {
            return null;
        }
        const storedReturnTo = popReturnTo(resolved.returnToStorageKey);
        const target = resolveReturnTarget({
            redirectTo: params?.redirectTo,
            returnTo: params?.returnTo,
            fallbackPath: params?.fallbackPath
        }, storedReturnTo, resolved);
        window.location.replace(target);
        return token;
    };
    return {
        options: resolved,
        createPkceBundle,
        createSignInUrl: (params) => createSignInUrl({
            shooBaseUrl: params.shooBaseUrl ?? resolved.shooBaseUrl,
            redirectUri: params.redirectUri ?? resolved.redirectUri,
            clientId: params.clientId ?? resolved.clientId,
            state: params.state,
            codeChallenge: params.codeChallenge,
            requestPii: params.requestPii ?? resolved.requestPii
        }),
        parseCallback,
        clearCallbackParams: () => {
            clearCallbackParams();
        },
        getIdentity: clientGetIdentity,
        persistIdentity: (userId, token, storageKey, extras) => {
            persistIdentity(userId, token, storageKey ?? resolved.storageKey, extras);
        },
        clearIdentity: (storageKey) => {
            clearIdentity(storageKey ?? resolved.storageKey);
        },
        decodeIdentityClaims,
        exchangeCode,
        checkSession: clientCheckSession,
        startSessionMonitor: clientStartSessionMonitor,
        startSignIn: clientStartSignIn,
        finishSignIn: clientFinishSignIn,
        handleCallback: clientHandleCallback
    };
}
